libraries/standard/http/src/http_client.c

```diff
@@ -1,42 +1,311 @@
+    }
+
+    /* Check if header value is long enough for length to overflow. */
+    if( valueLen > ( UINT32_MAX >> 2 ) )
+    {
+        /* TODO: Add log. */
+        status = HTTP_INVALID_PARAMETER;
+    }
+
+    /* "Content-Length" header must not be set by user if
+     * HTTP_REQUEST_DISABLE_CONTENT_LENGTH_FLAG is deactivated. */
+    if( strncmp( pField,
+                 HTTP_CONTENT_LENGTH_FIELD, HTTP_CONTENT_LENGTH_FIELD_LEN ) )
+    {
+        /* TODO: Add log. */
+        status = HTTP_INVALID_PARAMETER;
+    }
+
+    /* User must not set "Connection" header through this method. */
+    if( strncmp( pField,
+                 HTTP_CONNECTION_FIELD, HTTP_CONNECTION_FIELD_LEN ) )
+    {
+        /* TODO: Add log. */
+        status = HTTP_INVALID_PARAMETER;
+    }
+
+    /* User must not set "Host" header through this method. */
+    if( strncmp( pField,
+                 HTTP_HOST_FIELD, HTTP_HOST_FIELD_LEN ) )
+    {
+        /* TODO: Add log. */
+        status = HTTP_INVALID_PARAMETER;
+    }
+
+    /* User must not set "User-Agent" header through this method. */
+    if( strncmp( pField,
+                 HTTP_USER_AGENT_FIELD, HTTP_USER_AGENT_FIELD_LEN ) )
+    {
+        /* TODO: Add log. */
+        status = HTTP_INVALID_PARAMETER;
+    }
+
+    if( HTTP_SUCCEEDED( status ) )
+    {
+        status = _addHeader( pRequestHeaders,
+                             pField, fieldLen, pValue, valueLen );
+    }
+
+    return status;
 }
 
-HTTPStatus_t HTTPClient_Send( const HTTPTransportInterface_t* pTransport,
-                                  const HTTPRequestHeaders_t* pRequestHeaders,
-                                  const uint8_t* pRequestBodyBuf, // For a PUT or POST request.
-                                  size_t reqBodyBufLen,
-                                  httpResponse_t* pResponse )
+HTTPStatus_t HTTPClient_AddRangeHeader( HTTPRequestHeaders_t * pRequestHeaders,
+                                        int32_t rangeStart,
+                                        int32_t rangeEnd )
+{
+    /* Create buffer to fit max possible length for "bytes=<start>-<end>".
+     * This is the value of the Range header. */
+    char rangeValueStr[ HTTP_RANGE_BYTES_VALUE_MAX_LEN ] = { 0 };
+    char * pRangeValueCur = &rangeValueStr;
+    /* Excluding all the remaining null bytes. */
+    size_t rangeValueStrActualLength = 0;
+
+    /* Write "bytes=<start>:<end>" for Range header value. */
+    memcpy( rangeValueStr,
+            HTTP_RANGE_BYTES_PREFIX_VALUE, HTTP_RANGE_BYTES_PREFIX_VALUE_LEN );
+    pRangeValueCur += HTTP_RANGE_BYTES_PREFIX_VALUE_LEN;
+    rangeValueStrActualLength += HTTP_RANGE_BYTES_PREFIX_VALUE_LEN;
+    memcpy( rangeValueStr, EQUAL_CHARACTER, EQUAL_CHARACTER_LEN );
+    pRangeValueCur += EQUAL_CHARACTER_LEN;
+    rangeValueStrActualLength += EQUAL_CHARACTER_LEN;
+    memcpy( rangeValueStr, itoa( rangeStart ), itoaLength( rangeStart ) );
+    pRangeValueCur += itoaLength( rangeStart );
```

Comment:
@sarenameas I was unsure about using this itoaLength private method. Is this the best way to do this consider we can't use sprintf?

Reply:
we can use snprintf(), that is allowed and will return the number of bytes written (forgot if this include the null terminator you will have to look at the docs).